File: JavaGenerator/create_carsandtimes.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.HashSet;

/**
 * Created by Sung Kim on 3/2/2016.
 *
 * Create the carsandtimes after reading the single, clean, combined file.
 * The carsandtimes is simply each carid with its entry time, its exit time, and its xway.
 * This will be used by create_carstoreplace.java to create cars (by removing others) that re-enter the simulation on
 * a different xway, sometime later in the simulation.
 *
 * This is important as it prevents easy optimization via the assumption that a car's data can be localized to a single
 * xway-dir.
 *
 * We're only choosing cars that will _potentially_ be re-entrant based on a percentage of the total cars.
 * Thus, if the overlap factor is 10, then ~10% of the cars will be placed into the carsandtimes files.
 *
 * *********************************************************
 * This only creates the carsandtimes NOT the carstoreplace.
 * *********************************************************
 * java create_carsandtimes <combined file> <overlap factor: 0 - 100> <carsandtimes_outfile>
 */
public class create_carsandtimes {

    public static void main(String[] args) throws Exception {

        File infile = new File(args[0]);
        int overlap = Integer.parseInt(args[1]);
        File outfile = new File(args[2]);

        // Hold the [carid, ["Enter"|"Exit"|"Xway":value]].
        HashMap<Integer, HashMap<String, Integer>> cars = new HashMap<>();
        // Since we're choosing only a percentage of cars to be candidate cars.
        HashSet<Integer> rejects = new HashSet<>();

        BufferedReader reader = new BufferedReader(new FileReader(infile));
        String line;
        String[] tokens;
        int carid, time, xway;
        int r;
        PrintWriter writer = new PrintWriter(outfile);

        // Read th combined file and build out the cars and times (carid [entertime, exittime, xway]).
        while ((line = reader.readLine()) != null) {
            tokens = line.split(",");
            carid = Integer.parseInt(tokens[2]);
            time = Integer.parseInt(tokens[1]);
            xway = Integer.parseInt(tokens[4]);

            // Don't bother to check the cars that are already rejected.
            if (!rejects.contains(carid)) {
                // Update each exit time with the latest, if the car is not previously rejected.
                if (cars.containsKey(carid)) cars.get(carid).put("Exit", time);
                // A new car, never before seen. Check if this car will be a candiate.
                else {
                    r = (int) (Math.random() * 1000) % 100 + 1;
                    if (r < overlap) {
                        cars.put(carid, new HashMap<String, Integer>());
                        cars.get(carid).put("Enter", time);
                        cars.get(carid).put("Exit", time);
                        cars.get(carid).put("Xway", xway);
                    } else rejects.add(carid);
                }
            }
        }

        // For all the created candidate cars, print out the carsandtimes file.
        for (int cid : cars.keySet()) {
            writer.println(cid +","+cars.get(cid).get("Enter")+","+cars.get(cid).get("Exit")+","+cars.get(cid).get("Xway"));
        }
        writer.close();
    }
}

File: C/create_carsandtimes.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <time.h>

// Author: Sung Kim
// Create the carsandtimes after reading the single combined file
// ********************************************************
// This only creates the carsandtimes NOT the carstoreplace
// ********************************************************
// Usage: ./create_carsandtimes <combined file> <overlap factor: 0 - 100> <carsandtimes_outfile>

#define INITIAL_SIZE 100

// The structure to hold the relevant information for a car.
typedef struct carandtime
{
    int carid;
    int enter;
    int exit;
    int xway;
}
carandtime;

// Function to increase the size of an array on the heap.
void doublearr(int *arr, size_t size)
{
    int *temp = (int*) malloc(sizeof(int) * size * 2);
    memcpy(temp, arr, sizeof(int) * size);
    free(arr);
    arr = temp;
}

// Functions to turn an array of ints into a binary tree.
// Add.
int add(int *arr, int val, size_t size)
{

}

// Find.
int find(int *arr, int val, size_t size)
{
}

int main(int argc, char *argv[])
{

    char *infile = argv[0];
    int overlap = atoi(argv[1]);
    char *outfile = argv[2];

    // Normalize the rand()
    const int NORM_MAX = 100;
    int normalunit = RAND_MAX/NORM_MAX;

    // Hold the [carid, ["Enter"|"Exit"|"Xway":value]].
    // Use a struct for simplicity.
    carandtime *cars;
    // Since we're choosing only a percentage of cars to be candidate cars.
    int *rejects;

    size_t size = INITIAL_SIZE;
    // Start the sizes of the arrays with the default size.
    size_t cars_size = size;
    size_t rejects_size = size;

    cars = (carandtime*) malloc(sizeof(carandtime) * size);
    rejects = (int*) malloc(sizeof(int) * size);

    FILE *reader fopen(infile, "r");
    FILE *writer fopen(outfile, "w"); // Simple printf with redirection will be faster.

    // Hold the string array of tokens.
    char *token;
    // Hold the int array of tokens.
    // We only need carid (pos. 2), time (pos. 1), and xway (pos. 4) plus the others as placeholders, type (pos. 0)
    // and speed (pos. 3).
    int tokens[5];
    int carid, time, xway;
    int r;

    // Read th combined file and build out the cars and times (carid [entertime, exittime, xway]).
    // Open the file.
    FILE* fp = fopen(infile, "r");
    srand(time(NULL));
    // Ensure file exists or is otherwise readable.
   	if (fp != NULL)
   	{
   	    // Hold each line of the file as we read it.
        char line[100];
        // For building out the int representation of the array.
        int tokencount = 0;
        // Read the file line by line.
        while (fgets(line, 100, fp) != NULL)
        {
            // Tokenize the line.
            token = strtok(line, ",");
            // Convert the array of strings to ints.
            while (token != NULL || tokencount < 5)
            {
                tokens[tokencount] = atoi(token);
                // Move to the next token.
                token = strtok(NULL, ",");
                tokencount++;
            }
            // Reset the tokencount.
            tokencount = 0;

            carid = tokens[2];
            time = tokens[1];
            xway = tokens[4];

            // Don't bother to check the cars that are already rejected.
            if (find(rejects, carid) == -1)
            {
                // Update each exit time with the latest, if the car is not previously rejected.
                carandtime *temp = find(cars, carid);
                if (temp)
                {
                    temp->exit = time;
                }
                // A new car, never before seen. Check if this car will be a candiate.
                else
                {
                    if (rand() < normalunit * overlap)
                    {
                        carandtime *newcarandtime = (carandtime*) malloc(sizeof(carandtime));
                        newcarandtime->enter = time;
                        newcarandtime->exit = time;
                        newcarandtime->xway = xway;
                        add(cars, newcarandtime);
                    }
                    else
                    {
                        add(rejects, carid);
                    }
                }
            }
        }
    }

    // For all the created candidate cars, print out the carsandtimes file.
    for (int cid : cars.keySet())
    {
        writer.println(cid +","+cars.get(cid).get("Enter")+","+cars.get(cid).get("Exit")+","+cars.get(cid).get("Xway"));
    }
    writer.close();
}
